pwiz_tools/Shared/Common/Chemistry/MoleculeMassOffset.cs

```diff
@@ -29,73 +29,197 @@ namespace pwiz.Common.Chemistry
     /// Holds a chemical formula as well as an extra offset for the monoisotopic mass and the average mass.
     /// This is used when a molecule is made up of some things that we know the chemical formula for,
     /// and other things where the user has only told us the mono and average masses.
+    ///
+    /// Examples of these formulas with mass offsets include crosslinking and Hardklor output.
+    ///
     /// </summary>
-    public class MoleculeMassOffset : Immutable, IFormattable
+    public class MoleculeMassOffset : Immutable, IFormattable, IEquatable<MoleculeMassOffset>, IComparable<MoleculeMassOffset>
     {
-        public static readonly MoleculeMassOffset EMPTY = new MoleculeMassOffset(Molecule.Empty, 0, 0);
-        public MoleculeMassOffset(Molecule molecule, double monoMassOffset, double averageMassOffset)
+        public static readonly MoleculeMassOffset EMPTY = new MoleculeMassOffset(Molecule.Empty, TypedMass.ZERO_MONO_MASSNEUTRAL, TypedMass.ZERO_AVERAGE_MASSNEUTRAL);
+
+        protected MoleculeMassOffset(Molecule molecule, TypedMass monoMassOffset, TypedMass averageMassOffset) // Use Create instead of constructor to avoid creating empty objects
         {
-            Molecule = molecule;
+            Molecule = molecule ?? Molecule.Empty;
             MonoMassOffset = monoMassOffset;
             AverageMassOffset = averageMassOffset;
         }
 
-        public MoleculeMassOffset(Molecule molecule) : this(molecule, 0, 0)
+        public Molecule Molecule { get; private set; }
+        public TypedMass MonoMassOffset { get; protected set; }
```

Comment:
I am not sure I understand what effect changing this from a double to a TypedMass is supposed to do.
Are we trying to keep track of whether the mass is heavy, and/or is a massH?
What does it mean to add two massH's together? It looks like "operator +" in "TypedMass" has complicated rules about what to do when you the two operands differ in terms of heavy and massH.

Reply:
Yes, tracking the massH thing (for historical peptide reasons), and the Heavy (for small molecule mass-only description reasons).

Adding two massH together is fine, it's only important when interacting with protonated adducts to determine m/z. That's all historical accident and I wish we didn't need that distinction at all, but we do.